railgun/test/integration/knative_test.go

```diff
@@ -0,0 +1,286 @@
+//+build integration_tests
+
+package integration
+
+import (
+	"crypto/tls"
+	"net"
+	"net/http"
+	"os"
+	"os/exec"
+	"time"
+
+	"bytes"
+	"context"
+	"fmt"
+	"strings"
+	"testing"
+
+	types "k8s.io/apimachinery/pkg/types"
+
+	"github.com/kong/kubernetes-testing-framework/pkg/kind"
+	"github.com/stretchr/testify/assert"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	knativeCrds = "https://github.com/knative/serving/releases/download/v0.13.0/serving-crds.yaml"
+	knativeCore = "https://github.com/knative/serving/releases/download/v0.13.0/serving-core.yaml"
+)
+
+func isKnativeReady(ctx context.Context, cluster kind.Cluster) bool {
+	timeout := 1
+	for timeout < 60 {
+		podList, err := cluster.Client().CoreV1().Pods("knative-serving").List(ctx, metav1.ListOptions{})
+		if err != nil {
+			fmt.Println("failed retrieving knative pods.", err)
+			time.Sleep(1 * time.Second)
+			continue
+		}
+
+		ready := true
+		if len(podList.Items) != 4 {
+			fmt.Println("expected 4 pods, found ", len(podList.Items))
+			time.Sleep(1 * time.Second)
+			continue
+		}
+
+		for _, pod := range podList.Items {
+			if pod.Status.Phase != v1.PodRunning {
+				ready = false
+			}
+		}
+		if ready == true {
+			fmt.Println("All knative pods are up and ready.")
+			fmt.Println("Covering a window that webhook has been configured itself but not ready to receive traffic yet.")
+			time.Sleep(3 * time.Second)
+			return true
+		}
+		if ready == false {
+			time.Sleep(1 * time.Second)
+		}
+	}
+
+	fmt.Println("Failed to bring up knative resources within 60 seconds.")
+	return false
+}
+
+func TestKnativeIngress(t *testing.T) {
+	t.Skip("please use the steps run the test separately. ")
+	ctx := context.Background()
+
+	t.Log("Deploying all resources that are required to run knative")
+	err := deployManifest(knativeCrds, ctx)
+	assert.NoError(t, err)
+	err = deployManifest(knativeCore, ctx)
+	assert.NoError(t, err)
+	knativeReady := isKnativeReady(ctx, cluster)
+	assert.Equal(t, knativeReady, true)
+
+	t.Log("Note down the ip address or public CNAME of kong-proxy service.")
+	proxy, err := retrieveProxyInfo(ctx)
+	if err != nil {
+		t.Fatalf("kong-proxy service ip/public name is not ready.")
+	}
+
+	t.Log("Configure Knative NetworkLayer as Kong")
+	err = configKnativeNetwork(ctx, cluster)
+	assert.NoError(t, err)
+	err = configKnativeDomain(ctx, proxy, cluster)
+	assert.NoError(t, err)
+	// cover webhook sync windows
+	time.Sleep(3 * time.Second)
```

Comment:
This is technically a race condition, I recommend we use a `require.Eventually` with a specific API call which exercises the webhook to ensure it comes up before proceeding, but also fail here if the webhook does not provision properly.

Reply:
ATM, would say the knative webhook service readiness threshold is not stable - most cases ready for a post; sometimes, depends on the k8s cluster resources needs a window from configured to readiness. 
In our case,  I think a couple of seconds sync would be the most efficient way.

Reply:
I hope I can be persuasive here: the benefit of using `require.Eventually` is that you can perform some validating action and if this fails it will stop the entire test and report exactly where the problem occurred. The sleep method would be more prone to resulting in unknown behaviors: the later test calls will be written to expect that the hooks are working, but nothing actually validated that they did so the behavior of the tests past that point can be variable, resulting in some potential confusion under failure conditions. Let me know your thoughts?

Reply:
I hope following is clear: 
This unstable window after configuration before readiness is knative webhook service behavior, not Kong. As kong tests, we just use it.

Let me know what you think ?

Reply:
I'm still feeling pretty strongly that there's a more concise and definitive way we can test the "readiness" of the Knative webhook, rather than simply wait for a time period without validating that it's functional before we proceed to tests that assume they're functional. It seems like waiting, but then doing some kind of validating API call (like a dry run) would make for a stronger test.

Reply:
Agree.